File: boj/BFSDFS/BOJ_11123.py
# 백준 11123 Counting Sheep
# 21.01.22
import sys
sys.setrecursionlimit(10**6)


def DFS(r, c):
    if graph[r][c] == '.' or visit[r][c]: return
    visit[r][c] = True

    for i in range(4):
        nr = r + dy[i]
        nc = c + dx[i]

        if nr < 0 or nr >= n or nc < 0 or nc >= m: continue

        if graph[nr][nc] == '#' and not visit[nr][nc]:
            DFS(nr, nc)
    return


t = int(input())
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]
ans = []
while t:
    n, m = map(int, sys.stdin.readline().split())
    graph = [[] for _ in range(n)]
    visit = [[False] * m for _ in range(n)]

    for i in range(n):
        graph[i] = list(sys.stdin.readline().rstrip())

    cnt = 0
    for i in range(n):
        for j in range(m):
            if graph[i][j] == '#' and not visit[i][j]:
                DFS(i, j)
                cnt += 1

    ans.append(cnt)
    graph.clear()
    visit.clear()
    t -= 1

for i in ans:
    print(i)

File: boj/BFSDFS/BOJ_11123.cpp
/*----------------------
|백준 11123 Counting Sheep
|21.01.27
----------------------*/

#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
#include <string>

using namespace std;

int n, m;
int graph[101][101];
int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0, 0};

void DFS(int r, int c)
{
    if (graph[r][c] == 0) return;

    graph[r][c] = 0;

    for (int i = 0; i < 4; ++i)
    {
        int nr = r + dy[i];
        int nc = c + dx[i];

        if (nr < 0 || nr >= n || nc < 0 || nc >= m) continue;

        if (graph[nr][nc] == 1) DFS(nr, nc);
    }
    return;
}

int main()
{
    ios_base::sync_with_stdio(false); cin.tie(0);
    int t; cin >> t;

    vector <int> ans;

    while (t)
    {
        --t;

        cin >> n >> m;
        fill(&graph[0][0], &graph[100][101], 0);
        queue <pair <int, int> > q; // 양이 있는 위치
        int cnt = 0;

        for (int i = 0; i < n; ++i)
        {
            string s; cin >> s;

            for (int j = 0; j < s.length(); ++j)
            {
                if (s[j] == '#')
                {
                    graph[i][j] = 1;
                    q.push({i, j});
                }
            }
        }

        while (!q.empty())
        {
            int r = q.front().first;
            int c = q.front().second;
            q.pop();

            if (graph[r][c])
            {
                DFS(r, c);
                ++cnt;
            }
        }

        ans.push_back(cnt);
    }

    for (int i = 0; i < ans.size(); ++i)
        cout << ans[i] << "\n";

    return 0;
}